parlai/core/worlds.py

```diff
@@ -600,14 +600,14 @@ def batch_observe(self, index, batch_actions, index_acting):
             if batch_actions[i] is None:
                 # shouldn't send None, should send empty observations
                 batch_actions[i] = [{}] * len(self.worlds)
+            if index == index_acting:
+                return None  # don't observe yourself talking
             if hasattr(w, 'observe'):
                 # The world has its own observe function, which the action
                 # first goes through (agents receive messages via the world,
                 # not from each other).
                 observation = w.observe(agents[index], validate(batch_actions[i]))
             else:
-                if index == index_acting:
-                    return None  # don't observe yourself talking
```

Comment:
I think actually we don't want this code moved in case an ExecutableWorld wants to send you something based on your action (it could choose not to by doing this check there)